Feature / single page application

HmuuMyatMoe opened the pull request:
Added:
- Landing Page
- Dashboard Page for Admins (with CRUD operations to manage question repository)
- Question Page for Users (view questions operation)

Pughal77 committed:
Add build tool

Added gitignore as well

Pughal77 committed:
Add vite as dev dependency

Also created index.tsx in src as our index file for SPA

Pughal77 committed:
Add more dev dependencies

Vite is a fast build tool that offers hot reload for a better dev experience for frontend.

It also offers tree-shaking in which instead of importing a whole module that our code uses, only functions in that module that are used in the code will be imported.

Eslint and babel is also downloaded

This project will be written in typesript. With the type safety ts offers.we will be able to catch bugs earlier in development

Pughal77 committed:
Dev Environment for frontend setup

use `npm run start` to start the vite dev server to see the html output.
use `npm run build` to produce a build of the frontend

Pughal77 committed:
Update gitignore

Pughal77 committed:
Add tailwindcss

Tailwindcss allows us to design ui components with a faster development time as it comes with prebuilt CSS classes.

Pughal77 committed:
Add routing

Add the library `react-router-dom` to simplify the routing of our single-page-application

Pughal77 committed:
Add template pages

The Dashboard component - can leave the profile logo for later iterations

Pughal77 committed:
Rename dashboard to navbar

More appropriate name to its functionality

HmuuMyatMoe committed:
Setup tailwindcss

HmuuMyatMoe committed:
resolve merge conflicts

HmuuMyatMoe committed:
make AdminNavBar and Profile Button

Pughal77 committed:
Minor changes

Clearing unused imports

Pughal77 committed:
Update package-lock.json

HmuuMyatMoe committed:
Create Add question model

HmuuMyatMoe committed:
Create Edit question modal

HmuuMyatMoe committed:
Tidy up code

HmuuMyatMoe committed:
Create Delete Question Confirmation Modal

HmuuMyatMoe committed:
Fix bug in edit confirmation logic

Pughal77 committed:
Dasboard component

This allows us to display questions in our database to users

Pughal77 committed:
Merge branch 'feature/single-page-application' of https://github.com/CS3219-AY2425S1/cs3219-ay2425s1-project-g08 into feature/single-page-application

Pughal77 committed:
Uninstall axios

We realise that this library is redundant as we could use javascript's fetch api

HmuuMyatMoe committed:
change page nav

Pughal77 committed:
This is to allow CORS

Pughal77 committed:
Add httpclient5 dependency for allowing CORS

HmuuMyatMoe committed:
Merge branch 'dev' of https://github.com/CS3219-AY2425S1/cs3219-ay2425s1-project-g08 into feature/single-page-application

Pughal77 committed:
Allow CORS

Requests from localhost:5173 is accepted as that is where front end is being hosted

Pughal77 committed:
Reformat code

Pughal77 committed:
Merge branch 'feature/single-page-application' of https://github.com/CS3219-AY2425S1/cs3219-ay2425s1-project-g08 into feature/single-page-application

HmuuMyatMoe committed:
Merge branch 'feature/question-api' of https://github.com/CS3219-AY2425S1/cs3219-ay2425s1-project-g08 into feature/single-page-application

Pughal77 committed:
Incorporate edit Modal to dashboard

HmuuMyatMoe committed:
allow cors

HmuuMyatMoe committed:
Integrate API for add questions

HmuuMyatMoe committed:
resolve merge conflicts

HmuuMyatMoe committed:
fix add question api call

HmuuMyatMoe committed:
Integrate api for edit questions

HmuuMyatMoe committed:
Integrate api for delete questions

HmuuMyatMoe committed:
remove alerts

HmuuMyatMoe committed:
fix issue in delete question modal UI

HmuuMyatMoe committed:
Close modals only when add/edit are successful

HmuuMyatMoe committed:
Fix dropdown in edit confirmation modal

HmuuMyatMoe committed:
Change modal labels

Pughal77 committed:
format spacing of code

Pughal77 committed:
remove warning

text tag is not recognised by react, p\changed it to p tag

Pughal77 committed:
Merge branch 'feature/single-page-application' of https://github.com/CS3219-AY2425S1/cs3219-ay2425s1-project-g08 into feature/single-page-application

Pughal77 committed:
Update to check for repeated titles in repository

Pughal77 committed:
Edit the fetch function for editQuestion

DonovanJJ committed:
fix: Added temporary first to make the question dashboard list to rerender after adding new question

HmuuMyatMoe committed:
Merge branch 'feature/question-api' of https://github.com/CS3219-AY2425S1/cs3219-ay2425s1-project-g08 into feature/single-page-application

HmuuMyatMoe committed:
Add alerts for add/edit/delete fails

HmuuMyatMoe committed:
Edit alert messages

DonovanJJ committed:
added vscode file format to gitignore

DonovanJJ committed:
merge with upstream to fix edit button

HmuuMyatMoe committed:
Edit page nav

DonovanJJ committed:
fix: rerender question list when editing existing question

DonovanJJ committed:
fix: Rerender the page after deleting a question

HmuuMyatMoe committed:
Merge pull request #38 from CS3219-AY2425S1/fix/add-question-rerender

fix: Added temporary first to make the question dashboard list to rer…

DonovanJJ committed:
commted out title validation to fix issue regarding being unable to update the difficulty and category

HmuuMyatMoe committed:
Merge pull request #39 from CS3219-AY2425S1/fix/editing-question-difficulty

Commented out title validation to fix issue regarding being unable to u…

Pughal77 committed:
Create DasboardForUsers

This is to create a distinct workflow between admin and non-admin users

Pughal77 committed:
Merge branch 'feature/single-page-application' of https://github.com/CS3219-AY2425S1/cs3219-ay2425s1-project-g08 into feature/single-page-application

Pughal77 committed:
Refactor code

pass the function of fetch data to whichever component causes a change in questions state that is defined in the Dashboard component

Pughal77 committed:
Rough workflow for non-admin user

For a non-admin user, for now they would click the enter as user button, and then select a question they would like to view and that would bring them to the question page

HmuuMyatMoe committed:
Format code

HmuuMyatMoe committed:
Merge branch 'feature/single-page-application' of https://github.com/CS3219-AY2425S1/cs3219-ay2425s1-project-g08 into feature/single-page-application

HmuuMyatMoe committed:
comment out unused code

Pughal77 committed:
Fix Bug

Uncommenting this as this allows us to catch for the case of editing a question title to another question title that already exists. The frontend is also fixed to display the alert message that the title is repeated

Pughal77 committed:
Merge branch 'feature/single-page-application' of https://github.com/CS3219-AY2425S1/cs3219-ay2425s1-project-g08 into feature/single-page-application

Pughal77 committed:
Fix conflicts

HmuuMyatMoe committed:
Refactor and change to using States

HmuuMyatMoe committed:
Use Question to pass question information between components

HmuuMyatMoe committed:
Resolve tailwind css warning

HmuuMyatMoe committed:
Factor out DescriptionInput

HmuuMyatMoe committed:
Organise components

HmuuMyatMoe committed:
Resolve checkstyle warnings

DonovanJJ commented on frontend/src/components/Dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,128 @@
+
+// You can replace `any` with the actual type of questionList
+interface DashboardProps {
+  questions: Array<any>;
```
Since we are using Typescript, I'm thinking it might be beneficial if we enforce the type here as `Question` rather than `any`

Pughal77 replied:
Ah thank you, this will be enforced with the next commit

DonovanJJ commented on frontend/src/components/Dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,128 @@
+
+  const location = useLocation();
+
+  const onClick = (row: Row<{}>) => {
```
Very good usage of components to extract the displaying of questions as a table. 👍 
Just a random thought and up to your discretion.  Another possible implementation could be to pass the onClick callback function as a prop into this component, that way we need not check for pathname (No need to update two locations if we ever need it to change)

```
const DashBoard = func({ questions, fetchData, onClickCallback }) {
  return(
    {/* Pseudocode */}
    <Table>
      <Row onClick={() => onClickCallback(question)} />
    </Table>
  )
}
```

Pughal77 replied:
Ah yes that is a good suggestion. However, 

![image](https://github.com/user-attachments/assets/f870d4a2-d07e-4c34-8ece-04436834c177)

the onClick function above is using the openEditModal function that controls certain states within the Dashboard component that is not related to whether the user is admin or not. 

We are checking for the pathname  here as we need to check for if the user is admin or non-admin. Since the user service is not implemented yet, checking for the path name is meant to be a simple solution for the time being.

However, I believe that checking if the user is admin or not should be done in the component-level as this allows for code reusability (logic for what a user or admin can do is kept within the component level, so when creating the entire web page, we do not have to worry about passing in a prop to each and every component in the page for if the user is admin or non-admin)

DonovanJJ commented on frontend/src/components/NavBar.tsx:
Nit: It seems that we are mixing two different approaches of handling the navbar in this codebase.
1) Separate Navbars (AdminNavBar...)
2) Generic Navbar (handles the landingpage and user nav bar)
I understand that the navbars are quite similar but there are differences in logic between each navbar. 
However, I think it could be beneficial to just stick with one approach so that is becomes clearer as to what needs to change in the future.

So in this case, we could just stick with either:
- Separate navbars (LandingPage, Admin, User...) --> Sounds simpler since the navbar is going to differ between other pages (Dashboard, within text editor etc...)
- Just a generic navbar that allows for adding more links in

HmuuMyatMoe replied:
Thanks! We will look into both methods, especially the usage of a generic navbar that is reusable across different pages.

Pughal77 replied:
Yes I believe that we should work on creating a generic NavBar as well. This can be handled in the future!

DonovanJJ commented on frontend/src/components/QuestionModals/AddQuestionModal.tsx:
```diff
@@ -0,0 +1,185 @@
+  fetchData,
+  onClose,
+}) => {
+  const [complexityValue, setComplexityValue] = useState("");
```
Suggestion: Encapsulate all these into a single state - Furthermore, we could place these into a Type `FormValue` etc to maintain type safety

Sample code provided by chatgpt
```
type FormValues = {
  complexity: string;
  categories: string[];
  title: string;
  description: string;
};

const [formValues, setFormValues] = useState<FormValues>({
  complexity: "",
  categories: [""],
  title: "",
  description: "",
});

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { id, value } = e.target;

    setFormValues((prevValues) => ({
      ...prevValues,
      [id]: id === "categories" ? value.split(",").map((item) => item.trim()) : value,
    }));
  };
```

DonovanJJ commented on frontend/src/components/QuestionModals/EditQuestionModal.tsx:
```diff
@@ -0,0 +1,250 @@
+  const openEditConfirmationModal = () => setEditConfirmationModalOpen(true);
+  const closeEditConfirmationModal = () => setEditConfirmationModalOpen(false);
+
+  const [newComplexityValue, setNewComplexityValue] = useState(oldQuestion.complexity);
```
Likewise as mentioned in my previous comment, the state of the form could be encapsulated into a single state

DonovanJJ reviewed (approved):
The webpage looks really good! Thanks for the hard work 🙏 
Comments are mainly just suggestions, it's up to your discretion if you want to change them

I'll approve this PR so we can merge ASAP since functionality is already in place. Subsequent changes can be in a separate PR

Pughal77 committed:
Enforce Question type in Dashboard

Also refactored uesQuestion hook even further

Pughal77 committed:
Enforce Question in react table as well

Pughal77 committed:
Revert "Enforce Question in react table as well"

This reverts commit 9740bcfda7647c023cdebfcdc56fa3497fb9e53f.

Pughal77 committed:
Revert "Enforce Question type in Dashboard"

This reverts commit bede46cad874b1828ae6738d9a36d4fbe0b6a170.

Pughal77 committed:
Fix Uneccessary Render Bug

My previous 2 commits caused a huge bug that caused infinite renders. This is due to declaring my question state variable inside my useQuestions hook

Pughal77 committed:
Refactor without bug

When using a fetchData in a callback function for useEffect it is important to not include fetchData in the dependency array even if vs code prompts u to do so.

Including it in the dependency array would cause an infinite render loop.

Pughal77 reviewed (approved):
More Refactoring is to be done but since frontend is functional, we can push this for now.
